src/HashingLogic.test.ts

```diff
@@ -129,60 +181,275 @@ test('HashingLogic orderedStringify', () => {
+  expect(ethUtil.bufferToHex(oneRootChecksum).length).toBe(66)
+
+  const noRootChecksum = HashingLogic.getChecksum([])
+  expect(ethUtil.bufferToHex(noRootChecksum)).toBe(
+    '0x518674ab2b227e5f11e9084f615d57663cde47bce1ba168b4c19c7ee22a73d70'
+  )
+})
+
+test('HashingLogic.getSignedChecksum', () => {
+  const signedChecksumHash = HashingLogic.signChecksum(
+    fourRootHashes,
+    alicePrivkey
+  )
+  const sender = HashingLogic.recoverHashSigner(
+    HashingLogic.getChecksum(fourRootHashes),
+    signedChecksumHash
+  )
+
+  expect(sender.toLowerCase()).toBe(
+    aliceWallet.getAddressString().toLowerCase()
+  )
+  expect(sender.toLowerCase()).not.toBe(
+    bobWallet.getAddressString().toLowerCase()
   )
-  const phoneProof = tree.getProof(Buffer.from(hashedPhoneAttestation, 'hex'))
+})
+
+test('HashingLogic.hashAttestationTypes', () => {
+  // Deprecated
 
-  const treeDifferentOrder = HashingLogic.getMerkleTreeFromLeaves([
-    hashedEmailAttestation,
-    hashedFullNameAttestation,
-    hashedPhoneAttestation,
+  const emailAttestationTypeId = AttestationTypeID.email
+  const phoneAttestationTypeId = AttestationTypeID.phone
+  const emailAttestationTypeHash = HashingLogic.hashAttestationTypes([
+    emailAttestationTypeId,
+  ])
+  const phoneAttestationTypeHash = HashingLogic.hashAttestationTypes([
+    phoneAttestationTypeId,
   ])
+  const emailFirstCombinedAttestationTypeHash = HashingLogic.hashAttestationTypes(
+    [emailAttestationTypeId, phoneAttestationTypeId]
+  )
+  const phoneFirstCombinedAttestationTypeHash = HashingLogic.hashAttestationTypes(
+    [phoneAttestationTypeId, emailAttestationTypeId]
+  )
+
+  // Precomputed hashes should match (if they don't a bug has potentially been introduced)
+  expect(emailAttestationTypeHash).toBe(preComputedHashes.emailAttestationType)
+  expect(phoneAttestationTypeHash).toBe(preComputedHashes.phoneAttestationType)
+
+  // Differently sorted input arrays should output the same hash due to sorting prior to serialization / hashing
+  expect(emailFirstCombinedAttestationTypeHash).toBe(
+    phoneFirstCombinedAttestationTypeHash
+  )
+})
+
+test('HashingLogic.getPadding', () => {
+  // If 0 data node padding should be 0
+  let padding = HashingLogic.getPadding(0)
+  expect(padding.length).toBe(0)
 
-  const singleLeafTree = HashingLogic.getMerkleTree([emailAttestation])
-  const singleLeafTreeProof = singleLeafTree.getProof(
-    Buffer.from(hashedEmailAttestation, 'hex')
+  // If 1 data node padding should be 14
+  padding = HashingLogic.getPadding(1)
+  padding.forEach(p => expect(p.length).toBe(66))
+  expect(padding.length).toBe(14)
+
+  // If 10 data node padding should be 5
+  padding = HashingLogic.getPadding(10)
+  padding.forEach(p => expect(p.length).toBe(66))
+  expect(padding.length).toBe(5)
+
+  // If 15 data node padding should be 0
+  padding = HashingLogic.getPadding(15)
+  expect(padding.length).toBe(0)
+
+  // If 16 data node padding should be 495
```

Comment:
this is crazy, but do we actually foresee this happening in practice?

Reply:
Perhaps when we have super-meta-corporate attestations? might as well be prepared. probably won't see more than a few data nodes for a while